### DISCUSSION TITLE
关于feimi_shift的一些疑问

### DISCUSSION THREAD

**Discussion by hz-hz22** [hellozhaoming/FCP-vasp-ase#6] [Q&A]
老师您好，我当前用版本2进行恒电势计算，代码中是通过Ucal=-(self.fermi + self.fermishift)-self.workSHE来获得电位，请问式中-(self.fermi + self.fermishift)为什么会包含self.fermishift？

在另一篇文章中直接通过改nelect来改变费米能级，文中将电势设定为-0.48V，相应的费米能级直接修正到E-fermi :  -3.9485（来自outcar中的数据），并未考虑fermishift的影响（Comparative Study ofComputational Hydrogen Electrodes and Constant Electrode Potential Models Applied toElectrochemical Reduction ofCO 2 andOxygen Evolution Reaction onMetal Oxides/ Copper Catalysts），请问这里应该如何考虑。
![QQ截图20240902183126](https://github.com/user-attachments/assets/bdf1ce45-439e-493a-88b5-efe19bc2168b)

当前我在Ag111表面放了5个水分子和一个CO2分子，当存在显示水分子的时候，计算遇到了一些问题，在同一个电子数下对应多个能量值，请问这是什么原因，当存在显示水分子又开启了隐式溶剂化时，能给出详细的恒电势计算方法吗？
![QQ截图20240902193257](https://github.com/user-attachments/assets/0c22e67c-7de9-4784-9b5c-72e127f84422)
![能量](https://github.com/user-attachments/assets/da052ae8-a378-4c18-8a13-71f1a90a87ba)

另外，输出文件K的长度与log-fcp.txt不一致，这是什么原因？当进行频率计算和过渡态计算的时候是否也要加上电势？
期待您的回复

**Comment by hellozhaoming** [comment]
第一个问题，为什么会有个fermishift？因为VASP计算时晶胞的平均电势是零，如果真空层无限大，那么平均电势等于真空静电势，真空能级就是零。但是真空层不可能无限大，真空静电势就不是零。所以VASP中所有电子的能量参考点不是真空，需要进行一个能量矫正，这个能量矫正就是fermishift。。。如果没有做fermishift矫正，你就会发现巨正则能量的极小值点不是电势收敛的点，也就是恒电势优化出的结构不是能量极小值结构，可能会出现一些神奇但是不物理的性能（有时候神奇的计算结果投顶刊容易送审，不要迷信顶刊）。。。所以我建议所有的恒电势方法都要做一个改变电荷的静态计算，检查一下电势收敛的点是不是巨正则能量的极小值点。

第二个问题，一个电子数下对应多个能量，可能某一个状态不是巨正则的基态能量

第三个问题，显式水分子在隐式溶剂中无法正常形成氢键，氢键之间出现了不物理的隐式溶剂区域。如果不研究氢键或水团簇的影响应该问题不大，如果要研究氢键或水团簇，需要将隐式溶剂的区域限制在水分子层之外。可以试试gpaw的SJM模型或者vaspsol++。

第四个问题，k的输出有时是几个scf步拟合一个k，所以长度不一致，这个文件是用来调试用的。有些体系电子结构比较复杂，可能在不同的态之间跳跃，k的变化会比较大，此时建议手动调电子数

第五个问题，频率和过渡态计算都要加恒电势。所有的平衡态热力学或动力学量都需要电势收敛才能得到准确的能量。但不能用vasp自带的功能计算。建议用ASE中的频率和过渡态计算程序。如果过渡态或者震动过程电子数变化非常小，也可以直接固定电子数优化，后续再对过渡态和初态或末态进行单独的恒电势静态计算得到电子数，比较过渡态计算的电子数和恒电势计算得到的电子数，判断是不是真的变化特别小。可以根据巨正则的能量公式估算能量误差

**Reply by hz-hz22** [reply]
老师您好，十分感谢您的回复
还有一些问题想要请教您

1、进行恒电势计算时首先应该将结构优化到稳定状态再进行计算吗？正文里面写道_In addition, we focus here on the scheme in which ρ is optimized to the ground state at a constant Ne by the constant-charge SCF method and Ne is optimized in an outer loop to minimize Ω_.请问恒电势计算的时候可以同时优化结构吗？我的计算模型及计算脚本如下，Ag111表面吸附CO2分子底部两层固定，上面两层可动，恒电势计算过程中原子结构也在变化，但是在脚本里面并未指定NSW标签，请问这是正确的吗？

![image](https://github.com/user-attachments/assets/a7dc0dd0-4598-466c-bba7-f81782c062e5)
![图片1](https://github.com/user-attachments/assets/b9e14868-9548-43b2-95ca-5f25ae597029)

2、结构优化过程使用的INCAR设置及恒电势计算的输出如下，请问老师，是因为初始结构优化过程精度太差导致的恒电势计算过程比较慢吗？第一步FCP计算过程获得fmax值与基态计算之间存在较大差距。
ISTART=0
ICHARG=2
ISPIN=2
PREC=N
ALGO=N
NELM=80
EDIFF=1E-5
ENCUT=400
IVDW=11
IBRION=2
NSW=400
ISIF=2
EDIFFG=-0.05
ISMEAR=0
SIGMA=0.05
恒电势计算
![image](https://github.com/user-attachments/assets/fa1bd5db-96fa-461b-b423-f40b6ed31cad)

3、文章中方程1的解释中提到：Ne,PZC is the number of electrons at PZC. 也就是说Ne,PZC是基态计算时候的电子数，那么如何用FCP算法计算零电荷电势呢？我看很多是拟合出来的，在FCP算法中是不是可以设置电势进行计算，得到不同电势对应的电子数，直到计算结果中电子数收敛到Ne,PZC，那么此时的电势就是零电荷电势，请问老师可以这样理解吗？下图是其他文献中计算零电荷电势的方法
![image](https://github.com/user-attachments/assets/b3fc5f52-fa68-48f7-99bd-2eeaed0fd1d5)
4、FCP算法计算过程中3次循环就能将电势收敛到设定值，此时 conv=-0.000672，但体系的力一直 无法降至设定值0.01，进而导致结构及电子数都在优化，请问老师，这时候可以直接采用收敛的电子数直接进行结构优化计算吗，例如针对我的体系，目标电势为-0.8V，电子数为406.162849时，计算得到conv=-0.000672，此时，我可以固定电子数不变，在INCAR中设置NSW标签，之后不用FCP算法直接进行原子结构优化使得力降低到0.01ev吗？

刚开始学习DFT，问题比较多，请您多多包涵，期待您的回复，感谢老师提供的FCP算法，祝您科研顺利！！

**Reply by hellozhaoming** [reply]
1、进行恒电势计算时首先应该将结构优化到稳定状态再进行计算吗？不需要，但如果初始结构太离谱，可以先用vasp优化器做个粗优化。脚本里也不需要指定NSW。如果指定了NSW，则是用vasp的优化器，不是ase的优化器。

2、看起来你的恒电势计算并不慢，两个小时跑400步。你可以关闭vasp优化器，直接每一步都是恒电势优化。如果收敛还是慢，试试换一下优化器。在ase的官方文档里搜optimizer

3、如何用FCP算法计算零电荷电势呢？零电荷和恒电势矛盾，它是恒电荷，不是恒电势。直接把电子数设成vasp默认的电子数，直接跑vasp得到的结果就是零电荷的结果，此时的费米能级加上fermishif就是零电荷情况下的电势。

4、结构优化很难收敛时，建议增加vasp计算的精度和恒电势收敛的精度，以及使用不同的ase优化器。如果直接在INCAR中加NSW，得到的力的极小值点是恒电荷体系的极小值点，不一定是恒电势的极小值点
